libctru/source/services/fspxi.c

```diff
@@ -0,0 +1,1386 @@
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0xC, 4, 4);
+    cmdbuf[1] = (u32) file;
+    cmdbuf[2] = (u32)(file >> 32);
+    cmdbuf[3] = outSize;
+    cmdbuf[4] = inSize;
+    cmdbuf[5] = IPC_Desc_PXIBuffer(inSize, 0, true);
+    cmdbuf[6] = (u32) inBuffer;
+    cmdbuf[7] = IPC_Desc_PXIBuffer(outSize, 1, false);
+    cmdbuf[8] = (u32) outBuffer;
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_GetFileSize(Handle serviceHandle, FSPXI_File file, u64* size)
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0xD, 2, 0); // 0x000D0080
+    cmdbuf[1] = (u32) file;
+    cmdbuf[2] = (u32) (file >> 32);
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    if (size) *size = cmdbuf[2] | ((u64) cmdbuf[3] << 32);
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_SetFileSize(Handle serviceHandle, FSPXI_File file, u64 size)
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0xE, 4, 0); // 0x000E0100
+    cmdbuf[1] = (u32) size;
+    cmdbuf[2] = (u32) (size >> 32);
+    cmdbuf[3] = (u32) file;
+    cmdbuf[4] = (u32) (file >> 32);
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_CloseFile(Handle serviceHandle, FSPXI_File file)
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0xF, 2, 0); // 0x000F0080
+    cmdbuf[1] = (u32) file;
+    cmdbuf[2] = (u32) (file >> 32);
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_ReadDirectory(Handle serviceHandle, FSPXI_Directory directory, u32* entriesRead, u32 entryCount, FS_DirectoryEntry* entries)
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0x10, 3, 2); // 0x001000C2
+    cmdbuf[1] = (u32) directory;
+    cmdbuf[2] = (u32) (directory >> 32);
+    cmdbuf[3] = entryCount;
+    cmdbuf[4] = IPC_Desc_PXIBuffer(sizeof(FS_DirectoryEntry) * entryCount, 0, false);
+    cmdbuf[5] = (u32) entries;
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    if(entriesRead) *entriesRead = cmdbuf[2];
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_CloseDirectory(Handle serviceHandle, FSPXI_Directory directory)
+{
+    Result ret = 0;
+    u32* cmdbuf = getThreadCommandBuffer();
+
+    cmdbuf[0] = IPC_MakeHeader(0x11, 2, 0); // 0x00110080
+    cmdbuf[1] = (u32) directory;
+    cmdbuf[2] = (u32) (directory >> 32);
+
+    if(R_FAILED(ret = svcSendSyncRequest(serviceHandle))) return ret;
+
+    return (Result) cmdbuf[1];
+}
+
+Result FSPXI_OpenArchive(Handle serviceHandle, FSPXI_Archive* archive, FS_ArchiveID archiveID, FS_Path path)
+{
+    if (!archive) return -2;
```

Comment:
Shouldn't there be an `if (archive) ...` check below instead?

Reply:
This is in line with what FS does in FSUSER_OpenArchive. My guess is that it's a bit of added safety, that way users aren't opening up a bunch of archives without any way to close them later